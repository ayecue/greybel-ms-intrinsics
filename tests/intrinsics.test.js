const {
  Interpreter,
  Debugger,
  CustomFunction,
  OutputHandler,
  HandlerContainer,
  CustomString
} = require('greybel-interpreter');
const { init } = require('../dist');
const fs = require('fs');
const path = require('path');
const testFolder = path.resolve(__dirname, 'scripts');

let printMock;
const pseudoAPI = new Map();

global.Math.random = () => .5;

pseudoAPI.set(
  new CustomString('print'), 
  CustomFunction.createExternal(
    'print',
    (ctx, self, args) => {
      ctx.handler.outputHandler.print(ctx, args.get('value'));
    }
  ).addArgument('value')
);

class TestOutputHandler extends OutputHandler {
  print(_ctx, message) {
    printMock(message);
  }
}

class TestDebugger extends Debugger {
  debug() {}
}

describe('interpreter', function () {
  beforeEach(function () {
    printMock = jest.fn();
  });

  describe('default scripts', function () {
    fs.readdirSync(testFolder).forEach((file) => {
      const filepath = path.resolve(testFolder, file);

      test(path.basename(filepath), async () => {
        const interpreter = new Interpreter({
          target: filepath,
          api: init(pseudoAPI),
          handler: new HandlerContainer({
            outputHandler: new TestOutputHandler()
          }),
          debugger: new TestDebugger()
        });
        let success = false;

        try {
          await interpreter.run();
          success = true;
        } catch (e) {
          console.log(`${filepath} failed with: `, e);
        }

        expect(success).toEqual(true);
        for (const call of printMock.mock.calls) {
          expect(call[0]).toMatchSnapshot();
        }
      });
    });
  });
});
