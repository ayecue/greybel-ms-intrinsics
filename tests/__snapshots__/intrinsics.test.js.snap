// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpreter default scripts generics.src 1`] = `"\\"{\\""`;

exports[`interpreter default scripts generics.src 2`] = `"123"`;

exports[`interpreter default scripts generics.src 3`] = `"\\"0\\""`;

exports[`interpreter default scripts generics.src 4`] = `"1.23"`;

exports[`interpreter default scripts generics.src 5`] = `"0"`;

exports[`interpreter default scripts generics.src 6`] = `"101574"`;

exports[`interpreter default scripts generics.src 7`] = `"61655"`;

exports[`interpreter default scripts generics.src 8`] = `"61653"`;

exports[`interpreter default scripts generics.src 9`] = `"1"`;

exports[`interpreter default scripts generics.src 10`] = `"1345"`;

exports[`interpreter default scripts generics.src 11`] = `"[6, 5, 4, 3, 2, 1, 0]"`;

exports[`interpreter default scripts generics.src 12`] = `"[1, 2, 3, 4, 5]"`;

exports[`interpreter default scripts generics.src 13`] = `"[5, 4, 3, 2, 1]"`;

exports[`interpreter default scripts generics.src 14`] = `"[0, 2, 4, 6, 8, 10]"`;

exports[`interpreter default scripts generics.src 15`] = `"-1060040609"`;

exports[`interpreter default scripts generics.src 16`] = `"1.2055713536802566"`;

exports[`interpreter default scripts generics.src 17`] = `"0"`;

exports[`interpreter default scripts generics.src 18`] = `"1"`;

exports[`interpreter default scripts generics.src 19`] = `"1"`;

exports[`interpreter default scripts generics.src 20`] = `"null"`;

exports[`interpreter default scripts list.src 1`] = `"1"`;

exports[`interpreter default scripts list.src 2`] = `"0"`;

exports[`interpreter default scripts list.src 3`] = `"[0, 1]"`;

exports[`interpreter default scripts list.src 4`] = `"1"`;

exports[`interpreter default scripts list.src 5`] = `"null"`;

exports[`interpreter default scripts list.src 6`] = `"2"`;

exports[`interpreter default scripts list.src 7`] = `"\\"bar\\""`;

exports[`interpreter default scripts list.src 8`] = `"[\\"foo\\"]"`;

exports[`interpreter default scripts list.src 9`] = `"\\"foo\\""`;

exports[`interpreter default scripts list.src 10`] = `"[\\"bar\\"]"`;

exports[`interpreter default scripts list.src 11`] = `"[\\"foo\\", \\"bar\\", \\"xxx\\"]"`;

exports[`interpreter default scripts list.src 12`] = `"[1, 2, 4, 6]"`;

exports[`interpreter default scripts list.src 13`] = `"16"`;

exports[`interpreter default scripts list.src 14`] = `"null"`;

exports[`interpreter default scripts list.src 15`] = `"[\\"bar\\"]"`;

exports[`interpreter default scripts list.src 16`] = `"[\\"foo\\", \\"bar\\"]"`;

exports[`interpreter default scripts list.src 17`] = `"\\"foo;bar;xxx\\""`;

exports[`interpreter default scripts list.src 18`] = `"0"`;

exports[`interpreter default scripts list.src 19`] = `"0"`;

exports[`interpreter default scripts list.src 20`] = `"1"`;

exports[`interpreter default scripts list.src 21`] = `"1"`;

exports[`interpreter default scripts list.src 22`] = `"\\"true\\""`;

exports[`interpreter default scripts list.src 23`] = `"[{\\"foo\\":2}, {\\"foo\\":4}, {\\"foo\\":5}, {\\"foo\\":9}]"`;

exports[`interpreter default scripts list.src 24`] = `"[1, 2, 1337, 3, 4]"`;

exports[`interpreter default scripts list.src 25`] = `"[1, 3, 3, 7]"`;

exports[`interpreter default scripts manipulation.src 1`] = `"1"`;

exports[`interpreter default scripts manipulation.src 2`] = `"0"`;

exports[`interpreter default scripts manipulation.src 3`] = `"1"`;

exports[`interpreter default scripts manipulation.src 4`] = `"0"`;

exports[`interpreter default scripts manipulation.src 5`] = `"1"`;

exports[`interpreter default scripts manipulation.src 6`] = `"0"`;

exports[`interpreter default scripts manipulation.src 7`] = `"\\"lore\\""`;

exports[`interpreter default scripts manipulation.src 8`] = `"null"`;

exports[`interpreter default scripts manipulation.src 9`] = `"1"`;

exports[`interpreter default scripts manipulation.src 10`] = `"null"`;

exports[`interpreter default scripts manipulation.src 11`] = `"4"`;

exports[`interpreter default scripts manipulation.src 12`] = `"null"`;

exports[`interpreter default scripts manipulation.src 13`] = `"[\\"foo\\", \\"bar\\", \\"lore\\"]"`;

exports[`interpreter default scripts manipulation.src 14`] = `"[0, 1, 2]"`;

exports[`interpreter default scripts manipulation.src 15`] = `"[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]"`;

exports[`interpreter default scripts manipulation.src 16`] = `"[123, 1, \\"moo\\"]"`;

exports[`interpreter default scripts manipulation.src 17`] = `"[\\"xxx\\", 123, \\"foo\\"]"`;

exports[`interpreter default scripts manipulation.src 18`] = `"3"`;

exports[`interpreter default scripts manipulation.src 19`] = `"3"`;

exports[`interpreter default scripts manipulation.src 20`] = `"21"`;

exports[`interpreter default scripts manipulation.src 21`] = `"\\"abc lorem ipsum xxxav\\""`;

exports[`interpreter default scripts manipulation.src 22`] = `"\\"ABC LOREM IPSUM XXXAV\\""`;

exports[`interpreter default scripts manipulation.src 23`] = `"[123, \\"foo\\"]"`;

exports[`interpreter default scripts manipulation.src 24`] = `"[\\"xxx\\", 123]"`;

exports[`interpreter default scripts manipulation.src 25`] = `"\\" lorem ipsum XxXaV\\""`;

exports[`interpreter default scripts manipulation.src 26`] = `"\\"abc\\""`;

exports[`interpreter default scripts manipulation.src 27`] = `"124"`;

exports[`interpreter default scripts manipulation.src 28`] = `"123"`;

exports[`interpreter default scripts manipulation.src 29`] = `"0"`;

exports[`interpreter default scripts manipulation.src 30`] = `"123"`;

exports[`interpreter default scripts manipulation.src 31`] = `"\\"foo\\""`;

exports[`interpreter default scripts manipulation.src 32`] = `"{\\"bar\\":1,\\"lore\\":\\"moo\\"}"`;

exports[`interpreter default scripts manipulation.src 33`] = `"[\\"xxx\\", 123]"`;

exports[`interpreter default scripts manipulation.src 34`] = `"\\"xxx\\""`;

exports[`interpreter default scripts manipulation.src 35`] = `"{\\"bar\\":1,\\"lore\\":\\"moo\\"}"`;

exports[`interpreter default scripts manipulation.src 36`] = `"[123, \\"foo\\"]"`;

exports[`interpreter default scripts manipulation.src 37`] = `"{\\"foo\\":123,\\"bar\\":1,\\"lore\\":\\"moo\\",\\"addition\\":1}"`;

exports[`interpreter default scripts manipulation.src 38`] = `"[\\"xxx\\", 123, \\"foo\\", \\"addition\\"]"`;

exports[`interpreter default scripts manipulation.src 39`] = `"{\\"foo\\":123,\\"lore\\":\\"moo\\"}"`;

exports[`interpreter default scripts manipulation.src 40`] = `"[\\"xxx\\", \\"foo\\"]"`;

exports[`interpreter default scripts manipulation.src 41`] = `"\\"abc lorem  XxXaV\\""`;

exports[`interpreter default scripts manipulation.src 42`] = `"{\\"foo\\":123,\\"bar\\":\\"moo\\",\\"lore\\":1}"`;

exports[`interpreter default scripts manipulation.src 43`] = `"[\\"xxx\\", \\"foo\\", 123]"`;

exports[`interpreter default scripts manipulation.src 44`] = `"null"`;

exports[`interpreter default scripts manipulation.src 45`] = `"[\\"foo\\", \\"xxx\\", 123]"`;

exports[`interpreter default scripts manipulation.src 46`] = `"null"`;

exports[`interpreter default scripts manipulation.src 47`] = `"[123, \\"xxx\\", \\"foo\\"]"`;

exports[`interpreter default scripts map.src 1`] = `"1"`;

exports[`interpreter default scripts map.src 2`] = `"0"`;

exports[`interpreter default scripts map.src 3`] = `"[\\"foo\\", \\"bar\\"]"`;

exports[`interpreter default scripts map.src 4`] = `"[9064, \\"ipsum\\"]"`;

exports[`interpreter default scripts map.src 5`] = `"\\"bar\\""`;

exports[`interpreter default scripts map.src 6`] = `"null"`;

exports[`interpreter default scripts map.src 7`] = `"2"`;

exports[`interpreter default scripts map.src 8`] = `"123"`;

exports[`interpreter default scripts map.src 9`] = `"{\\"bar\\":\\"ipsum\\"}"`;

exports[`interpreter default scripts map.src 10`] = `"{\\"foo\\":123,\\"bar\\":\\"ipsum\\",\\"moo\\":1}"`;

exports[`interpreter default scripts map.src 11`] = `"323"`;

exports[`interpreter default scripts map.src 12`] = `"{\\"foo\\":123,\\"bar\\":\\"ipsum\\"}"`;

exports[`interpreter default scripts map.src 13`] = `"1"`;

exports[`interpreter default scripts map.src 14`] = `"{\\"test\\":3,\\"bar\\":3,\\"foo\\":1}"`;

exports[`interpreter default scripts math.src 1`] = `"1337"`;

exports[`interpreter default scripts math.src 2`] = `"NaN"`;

exports[`interpreter default scripts math.src 3`] = `"NaN"`;

exports[`interpreter default scripts math.src 4`] = `"1.5700483837780477"`;

exports[`interpreter default scripts math.src 5`] = `"-3.878660859970253"`;

exports[`interpreter default scripts math.src 6`] = `"1337"`;

exports[`interpreter default scripts math.src 7`] = `"0.2496568790410366"`;

exports[`interpreter default scripts math.src 8`] = `"1337"`;

exports[`interpreter default scripts math.src 9`] = `"-0.9683343651587963"`;

exports[`interpreter default scripts math.src 10`] = `"1337"`;

exports[`interpreter default scripts math.src 11`] = `"36.565010597564445"`;

exports[`interpreter default scripts math.src 12`] = `"3.141592653589793"`;

exports[`interpreter default scripts math.src 13`] = `"281"`;

exports[`interpreter default scripts math.src 14`] = `"1403"`;

exports[`interpreter default scripts math.src 15`] = `"1122"`;

exports[`interpreter default scripts math.src 16`] = `"-939524096"`;

exports[`interpreter default scripts math.src 17`] = `"0"`;

exports[`interpreter default scripts math.src 18`] = `"0"`;

exports[`interpreter default scripts math.src 19`] = `"-1338"`;

exports[`interpreter default scripts rnd.src 1`] = `"0.4985665436834097"`;

exports[`interpreter default scripts rnd.src 2`] = `"0.04626259859651327"`;

exports[`interpreter default scripts rnd.src 3`] = `"0.04173178947530687"`;

exports[`interpreter default scripts rnd.src 4`] = `"0.22427810914814472"`;

exports[`interpreter default scripts rnd.src 5`] = `"0.2948848782107234"`;

exports[`interpreter default scripts rnd.src 6`] = `"0.8563546116929501"`;

exports[`interpreter default scripts string.src 1`] = `"1"`;

exports[`interpreter default scripts string.src 2`] = `"0"`;

exports[`interpreter default scripts string.src 3`] = `"[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]"`;

exports[`interpreter default scripts string.src 4`] = `"[\\"x\\", \\"x\\", \\"x\\", \\" \\", \\"l\\", \\"o\\", \\"r\\", \\"e\\", \\"m\\", \\" \\", \\"i\\", \\"p\\", \\"s\\", \\"u\\", \\"m\\", \\" \\", \\"b\\", \\"a\\", \\"r\\", \\" \\", \\"f\\", \\"o\\", \\"r\\"]"`;

exports[`interpreter default scripts string.src 5`] = `"16"`;

exports[`interpreter default scripts string.src 6`] = `"null"`;

exports[`interpreter default scripts string.src 7`] = `"120"`;

exports[`interpreter default scripts string.src 8`] = `"23"`;

exports[`interpreter default scripts string.src 9`] = `"\\"xxx was ccccccc\\""`;

exports[`interpreter default scripts string.src 10`] = `"1.23"`;

exports[`interpreter default scripts string.src 11`] = `"0"`;

exports[`interpreter default scripts string.src 12`] = `"\\"XXX  CCCCCCC\\""`;

exports[`interpreter default scripts string.src 13`] = `"\\"XXX WAS CCCCCCC\\""`;

exports[`interpreter default scripts string.src 14`] = `"[\\"123\\", \\"was\\", \\"false\\"]"`;

exports[`interpreter default scripts string.src 15`] = `"\\"xxx replaced ipsum bar for\\""`;

exports[`interpreter default scripts string.src 16`] = `"\\"was      xx\\""`;

exports[`interpreter default scripts string.src 17`] = `"\\"1213373456\\""`;

exports[`interpreter default scripts string.src 18`] = `"[\\"a\\", \\"b\\", \\"c\\", \\"e\\", \\"d\\"]"`;

exports[`interpreter default scripts string.src 19`] = `"[\\"abced\\", \\"ca\\", \\"xa\\"]"`;

exports[`interpreter default scripts string.src 20`] = `"\\"foo \\\\\\"bar\\\\\\" foo\\""`;

exports[`interpreter default scripts string.src 21`] = `"\\"foo \\"bar\\" foo\\""`;

exports[`interpreter default scripts string.src 22`] = `"[\\"ab.ce.d\\"]"`;

exports[`interpreter default scripts string.src 23`] = `"\\"xxx lrem ipsum bar fr\\""`;
